--- libavcodec/options.c	2007-06-07 16:05:43.000000000 +0100
+++ libavcodec/options.c	2007-06-07 16:05:12.000000000 +0100
@@ -423,7 +423,7 @@
 #define E AV_OPT_FLAG_ENCODING_PARAM
 #define D AV_OPT_FLAG_DECODING_PARAM
 
-#define AV_CODEC_DEFAULT_BITRATE 200*1000
+#define AV_CODEC_DEFAULT_BITRATE 1100*1000
 
 static const AVOption options[]={
 {"b", "set video bitrate (in bits/s)", OFFSET(bit_rate), FF_OPT_TYPE_INT, AV_CODEC_DEFAULT_BITRATE, INT_MIN, INT_MAX, V|E},
