diff -r -u libavformat/movenc.c ../ffmpeg-0.5-jf/libavformat/movenc.c
--- libavformat/movenc.c	2009-02-28 16:02:29.000000000 +0000
+++ ../ffmpeg-0.5-jf/libavformat/movenc.c	2009-04-17 15:20:19.000000000 +0100
@@ -596,12 +596,19 @@
                         (tag == MKTAG('d','v','c','p') ||
                          track->enc->codec_id == CODEC_ID_RAWVIDEO))) {
         if (track->enc->codec_id == CODEC_ID_DVVIDEO) {
-            if (track->enc->height == 480) /* NTSC */
-                if  (track->enc->pix_fmt == PIX_FMT_YUV422P) tag = MKTAG('d','v','5','n');
-                else                                         tag = MKTAG('d','v','c',' ');
-            else if (track->enc->pix_fmt == PIX_FMT_YUV422P) tag = MKTAG('d','v','5','p');
-            else if (track->enc->pix_fmt == PIX_FMT_YUV420P) tag = MKTAG('d','v','c','p');
-            else                                             tag = MKTAG('d','v','p','p');
+            if (track->enc->width == 720) /* SD */
+                if (track->enc->height == 480) /* NTSC */
+                    if  (track->enc->pix_fmt == PIX_FMT_YUV422P) tag = MKTAG('d','v','5','n');
+                    else                                         tag = MKTAG('d','v','c',' ');
+                else if (track->enc->pix_fmt == PIX_FMT_YUV422P) tag = MKTAG('d','v','5','p');
+                else if (track->enc->pix_fmt == PIX_FMT_YUV420P) tag = MKTAG('d','v','c','p');
+                else                                             tag = MKTAG('d','v','p','p');
+            else if (track->enc->height == 720) /* HD 720 line */
+                if  (track->enc->time_base.den == 50)            tag = MKTAG('d','v','h','q');
+                else                                             tag = MKTAG('d','v','h','p');
+            else /* HD 1080 line */
+                if  (track->enc->time_base.den == 25)            tag = MKTAG('d','v','h','5');
+                else                                             tag = MKTAG('d','v','h','6');
         } else if (track->enc->codec_id == CODEC_ID_RAWVIDEO) {
             tag = codec_get_tag(mov_pix_fmt_tags, track->enc->pix_fmt);
             if (!tag) // restore tag
