--- ./libavcodec/dnxhddec.c	2009-01-19 15:46:40.000000000 +0000
+++ ../ffmpeg-0.5/libavcodec/dnxhddec.c	2011-07-05 11:00:25.000000000 +0100
@@ -126,12 +126,17 @@
     ctx->mb_width = ctx->width>>4;
     ctx->mb_height = buf[0x16d];
 
-    if (ctx->mb_height > 68) {
-        av_log(ctx->avctx, AV_LOG_ERROR, "mb height too big\n");
+    dprintf(ctx->avctx, "mb width %d, mb height %d\n", ctx->mb_width, ctx->mb_height);
+
+    if ((ctx->height+15)>>4 == ctx->mb_height && ctx->picture.interlaced_frame)
+        ctx->height <<= 1;
+
+    if (ctx->mb_height > 68 ||
+        (ctx->mb_height<<ctx->picture.interlaced_frame) > (ctx->height+15)>>4) {
+        av_log(ctx->avctx, AV_LOG_ERROR, "mb height too big: %d\n", ctx->mb_height);
         return -1;
     }
 
-    dprintf(ctx->avctx, "mb width %d, mb height %d\n", ctx->mb_width, ctx->mb_height);
     for (i = 0; i < ctx->mb_height; i++) {
         ctx->mb_scan_index[i] = AV_RB32(buf + 0x170 + (i<<2));
         dprintf(ctx->avctx, "mb scan index %d\n", ctx->mb_scan_index[i]);
@@ -290,6 +295,13 @@
     if (dnxhd_decode_header(ctx, buf, buf_size, first_field) < 0)
         return -1;
 
+    if ((avctx->width || avctx->height) &&
+        (ctx->width != avctx->width || ctx->height != avctx->height)) {
+        av_log(avctx, AV_LOG_WARNING, "frame size changed: %dx%d -> %dx%d\n",
+               avctx->width, avctx->height, ctx->width, ctx->height);
+        first_field = 1;
+    }
+
     avctx->pix_fmt = PIX_FMT_YUV422P;
     if (avcodec_check_dimensions(avctx, ctx->width, ctx->height))
         return -1;
