--- libavformat/movenc.c	2009-02-28 16:02:29.000000000 +0000
+++ ../ffmpeg-0.5-aspect-ratio/libavformat/movenc.c	2010-06-14 14:38:29.000000000 +0100
@@ -69,6 +69,7 @@
     uint8_t     *vosData;
     MOVIentry   *cluster;
     int         audio_vbr;
+    int         width;  ///< visual width for DVCPRO-HD
     int         height; ///< active picture (w/o VBI) height for D-10/IMX
     int         tref;
 } MOVTrack;
@@ -664,6 +665,15 @@
     return updateSize(pb, pos);
 }
 
+static int mov_write_pasp_tag(ByteIOContext *pb, MOVTrack *track)
+{
+    put_be32(pb, 16);
+    put_tag(pb, "pasp");
+    put_be32(pb, track->enc->sample_aspect_ratio.num);
+    put_be32(pb, track->enc->sample_aspect_ratio.den);
+    return 16;
+}
+
 static int mov_write_video_tag(ByteIOContext *pb, MOVTrack *track)
 {
     int64_t pos = url_ftell(pb);
@@ -725,6 +735,9 @@
     } else if(track->vosLen > 0)
         mov_write_glbl_tag(pb, track);
 
+    if (track->enc->sample_aspect_ratio.num)
+        mov_write_pasp_tag(pb, track);
+
     return updateSize(pb, pos);
 }
 
@@ -1046,10 +1059,7 @@
     /* Track width and height, for visual only */
     if(track->enc->codec_type == CODEC_TYPE_VIDEO ||
        track->enc->codec_type == CODEC_TYPE_SUBTITLE) {
-        double sample_aspect_ratio = av_q2d(st->sample_aspect_ratio);
-        if(!sample_aspect_ratio || track->height != track->enc->height)
-            sample_aspect_ratio = 1;
-        put_be32(pb, sample_aspect_ratio * track->enc->width*0x10000);
+        put_be32(pb, track->width*0x10000);
         put_be32(pb, track->height*0x10000);
     }
     else {
@@ -1683,6 +1693,28 @@
                 track->height = track->tag>>24 == 'n' ? 486 : 576;
             } else
                 track->height = st->codec->height;
+
+            if (track->tag == MKTAG('d','v','h','5')) {
+                if (st->codec->width != 1440) {
+                    av_log(s, AV_LOG_ERROR, "DVCPro-HD 1080/50i width must be 1440\n");
+                    return -1;
+                }
+                track->width = 1920;
+            } else if (track->tag == MKTAG('d','v','h','6')) {
+                if (st->codec->width != 1280) {
+                    av_log(s, AV_LOG_ERROR, "DVCPro-HD 1080/60i width must be 1280\n");
+                    return -1;
+                }
+                track->width = 1920;
+            } else if (track->tag == MKTAG('d','v','h','q') || track->tag == MKTAG('d','v','h','p')) {
+                if (st->codec->width != 960) {
+                    av_log(s, AV_LOG_ERROR, "DVCPro-HD 720/60p or 720/50p width must be 960\n");
+                    return -1;
+                }
+                track->width = 1280;
+            } else
+                track->width = st->codec->width;
+
             track->timescale = st->codec->time_base.den;
             av_set_pts_info(st, 64, 1, st->codec->time_base.den);
             if (track->mode == MODE_MOV && track->timescale > 100000)
